internal/balancergroup/balancergroup_test.go

```diff
@@ -534,3 +534,118 @@ func (s) TestBalancerExitIdleOne(t *testing.T) {
+	bg.UpdateClientConnState(testBalancerIDs[0], balancer.ClientConnState{ResolverState: resolver.State{Addresses: testBackendAddrs[0:2]}})
+
+	bg.Start()
+
+	m1 := make(map[resolver.Address]balancer.SubConn)
+	for i := 0; i < 2; i++ {
+		addrs := <-cc.NewSubConnAddrsCh
+		sc := <-cc.NewSubConnCh
+		m1[addrs[0]] = sc
+		bg.UpdateSubConnState(sc, balancer.SubConnState{ConnectivityState: connectivity.Connecting})
+		bg.UpdateSubConnState(sc, balancer.SubConnState{ConnectivityState: connectivity.Ready})
+	}
+
+	// Should behave like a normal rr picker
+	p1 := <-cc.NewPickerCh
+	want := []balancer.SubConn{
+		m1[testBackendAddrs[0]], m1[testBackendAddrs[1]],
+	}
+	if err := testutils.IsRoundRobin(want, subConnFromPicker(p1)); err != nil {
+		t.Fatalf("want %v, got %v", want, err)
+	}
+
+	// The balancer type for testBalancersIDs[0] is currently Round Robin. Now,
+	// change it to a balancer that has separate behavior logically (creating
+	// SubConn for second address in address list and always picking that
+	// SubConn), and see if the aggregated picker reflects that change.
+	bg.UpdateBuilder(testBalancerIDs[0], secondAddressBalancerBuilder{})
+	if err := bg.UpdateClientConnState(testBalancerIDs[0], balancer.ClientConnState{ResolverState: resolver.State{Addresses: testBackendAddrs[2:4]}}); err != nil {
+		t.Fatalf("error updating ClientConn state: %v", err)
+	}
+
+	addrs := <-cc.NewSubConnAddrsCh
+	if addrs[0].Addr != testBackendAddrs[3].Addr {
+		t.Fatalf("newSubConn called with wrong address, want: %v, got : %v", testBackendAddrs[3].Addr, addrs[0].Addr)
+	}
+	sc := <-cc.NewSubConnCh
+
+	// Pending being ready should forward the new picker (wrapped in a weighted
+	// aggregator) to the ClientConn which should only return SubConn
+	// corresponding to address 3.
+	p2 := <-cc.NewPickerCh
+	want = []balancer.SubConn{
+		sc,
+	}
+	if err := testutils.IsRoundRobin(want, subConnFromPicker(p2)); err != nil {
+		t.Fatalf("want %v, got %v", want, err)
+	}
+}
+
+type secondAddressBalancerBuilder struct{}
+
+func (secondAddressBalancerBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	return &secondAddressBalancer{
+		cc: cc,
+	}
+}
+
+func (secondAddressBalancerBuilder) Name() string {
+	return "secondAddressBalancer"
+}
+
+type secondAddressBalancer struct {
+	cc balancer.ClientConn
+	p  balancer.Picker
+}
+
+func (mb *secondAddressBalancer) UpdateClientConnState(ccs balancer.ClientConnState) error {
+	// Hardcode this balancer to only deal with the second address to verify
+	// that it's this balancer type in the balancer group.
+	sc, err := mb.cc.NewSubConn(ccs.ResolverState.Addresses[1:2], balancer.NewSubConnOptions{})
+	if err != nil {
+		return err
+	}
+
+	mb.p = &secondAddressBalancerPicker{
+		sc: sc,
+	}
+
+	// READY will move this balancer into current, causing it to Update Client
+	// Conn with new Picker for this specific type of balancer, not round robin.
+	mb.cc.UpdateState(balancer.State{
```

Comment:
Ideally, this should only happen after the SubConn state turns Ready.

And for this testing balancer, you can wrap pickfirst https://github.com/grpc/grpc-go/blob/master/pickfirst.go#L30
And override `UpdateClientConnState` to always delete the first address

Reply:
Switched to a wrapped PickFirst. I had to intercept an UpdateState() call with Idle as it calls that inline on the first successful UpdateState call for this to only trigger on SubConn being Ready. This also allowed me to verify that the Picker was not updated until Graceful Switch process was complete, as I now have a knob on that due to it being triggered by SubConnState change.